python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -896,106 +871,128 @@ class PyInferenceRequest
+        nullptr /* start_fn */));
+    allocator_.reset(allocator_raw, TRITONSERVER_ResponseAllocatorDelete);
+    ThrowIfError(TRITONSERVER_ResponseAllocatorSetQueryFunction(
+        allocator_.get(), ResponseAllocatorQueryFn));
+    ThrowIfError(TRITONSERVER_ResponseAllocatorSetBufferAttributesFunction(
+        allocator_.get(), ResponseAllocatorBufferAttributesFn));
+  }
+  static TRITONSERVER_Error* ResponseAllocatorQueryFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* userp,
+      const char* tensor_name, size_t* byte_size,
+      TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id)
+  {
+    *memory_type = TRITONSERVER_MEMORY_CPU;
+    *memory_type_id = 0;
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorBufferAttributesFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      struct TRITONSERVER_BufferAttributes* buffer_attributes, void* userp,
+      void* buffer_userp)
   {
-    allocator_callback_resource_.reset(
-        new PyResponseAllocator::CallbackResource(
-            allocator, allocater_user_object));
-    response_callback_resource_.reset(new PyInferenceResponse::CallbackResource(
-        response, allocator_callback_resource_.get(), response_user_object));
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryType(
+        buffer_attributes, TRITONSERVER_MEMORY_CPU));
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryTypeId(
+        buffer_attributes, 0 /* memory_type_id */));
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorAllocFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      size_t byte_size, TRITONSERVER_MemoryType memory_type,
+      int64_t memory_type_id, void* userp, void** buffer, void** buffer_userp,
+      TRITONSERVER_MemoryType* actual_memory_type,
+      int64_t* actual_memory_type_id)
+  {
+    *buffer = malloc(byte_size * sizeof(uint8_t));
+    *actual_memory_type = TRITONSERVER_MEMORY_CPU;
+    *actual_memory_type_id = 0;
+    // The response allocator needs to be kept alive until the allocated memory
+    // is released via the release function.
+    std::unique_ptr<std::shared_ptr<struct TRITONSERVER_ResponseAllocator>>
+        allocator_ptr(
+            new std::shared_ptr<struct TRITONSERVER_ResponseAllocator>(
+                *reinterpret_cast<
+                    std::shared_ptr<struct TRITONSERVER_ResponseAllocator>*>(
+                    userp)));
+    *buffer_userp = reinterpret_cast<void*>(allocator_ptr.release());
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorReleaseFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* buffer,
+      void* buffer_userp, size_t byte_size, TRITONSERVER_MemoryType memory_type,
+      int64_t memory_type_id)
+  {
+    free(buffer);
+    // Release ownership of the response allocator.
```

Comment:
thinking a singleton could be simpler and avoid alloc / dealloc of the allocator object itself - if possible.

Reply:
yes, updated to use a singleton response allocator.